File: Tree/q246_inorder_successor_in_bst.java
package Tree;

//    Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class q246_inorder_successor_in_bst {
    // 法1：
    public TreeNode inorderSuccessor1(TreeNode root, TreeNode p) {
        TreeNode successor = null;
        // 退出：未找到p/root为空  || 找到p
        while (root != null && root.val != p.val) {
            if (root.val > p.val) { // 左子树中找
                // 0.[左 [根] 右], 找到p:
                // 1.若successor无右子树，即可直接返回
                // 2.若有右子树，则返回右子树的最左孩子
                successor = root;
                root = root.left;
            } else { // if root.val < p.val[不可能等于 因为等于时会跳出]
                root = root.right;
            }
        }

        if (root == null) return null;
        if (root.right == null) return successor; // 1.
        // 2. 找到右子树的最左孩子
        root = root.right;
        while (root.left != null) {
            root = root.left;
        }
        return root;
    }

        // 法2：【荐！】
    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        TreeNode res = null;
        while (root != null) {
            if (root.val > p.val) {
                res = root; // 1.左子树的根 2.右子树的最左孩子
                root = root.left;
            } else // <= 在右子树中找
                root = root.right;
        }
        return res;
    }

        // 法3：
    public TreeNode inorderSuccessor3(TreeNode root, TreeNode p) {
        if (root == null || p == null) return null;
        if (root.val <= p.val) { // <=  相等时找右子树，∵左根[右]
            // 若右子树中（下面的else语句）：1.无左子树 直接返回右子树的根 2.否则返回右子树的最左孩子
            return inorderSuccessor3(root.right, p);
        } else { // >  0. 左[根]右，若[根]无右孩子 直接返回，否则继续找右子树中的最左孩子(上面的if语句)
            TreeNode leftSuc = inorderSuccessor3(root.left, p);
            return (leftSuc != null)? leftSuc : root;
        }
    }
}

File: DataStructure/Tree/q246_inorder_successor_in_bst.py
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None


class TreeNode2(object):
    def __init__(self, x):
        self.val = x
        self.left, self.right = None, None
        self.father = None


class Solution:
    def inorderSuccessor(self, root: TreeNode, p: TreeNode) -> TreeNode:
        successor = None
        while root and root.val != p.val:
            if root.val > p.val:
                successor = root
                root = root.left
            else:
                root = root.right

        if not root: return None
        if not root.right: return successor

        # 有右子树，则找右子树的最左孩子
        root = root.right
        while root.left:
            root = root.left
        return root

    # 变题：若题目不给root，只给出结点，增加结点的'父节点'属性：
    def inorderSuccessor2(self, q: TreeNode2) -> TreeNode2:
        if q.right:
            q = q.right
            while q.left:
                q = q.left
            return q
        # q无右子树，则返回某祖宗结点res，其中res的左子树中包含q（最近祖先）
        # 找出q祖宗的第一个拐点(整体作为祖宗的左子树)
        while q.father:
            if q.father.left == q:
                return q.father
            else:
                q = q.father
